llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -109,8 +201,50 @@ SYCLVirtualFunctionsAnalysisPass::run(Module &M, ModuleAnalysisManager &MAM) {
   }
 
   // Emit a diagnostic if a kernel performs virtual function calls
-  for (const auto *K : Kernels) {
+  for (auto *K : KernelsToCheck)
     checkKernel(K, CallGraph);
+
+  // Cache to know which function is used by which kernels
+  FuncToFuncMapTy FunctionToKernels;
+  for (auto *K : AllKernels)
+    computeFunctionToKernelsMapping(K, CallGraph, FunctionToKernels);
+
+  for (const auto *F : IndirectlyCallableFuncs) {
+    StringRef Set = F->getFnAttribute("indirectly-callable").getValueAsString();
+
+    SmallVector<const GlobalVariable *, 4> VTables;
+    collectVTablesThatUseFunction(F, VTables);
+    SmallPtrSet<Function *, 8> KernelsToUpdate;
+
+    for (const auto *GV : VTables) {
+      // Find functions that use those vtables
+      SmallPtrSet<const Function *, 16> FunctionsThatUseVTables;
+      for (const auto *UU : GV->users())
+        collectEnclosingFunctions(UU, FunctionsThatUseVTables);
+      // And collect kernels that use those functions
+      for (const Function *FF : FunctionsThatUseVTables)
+        for (auto *K : FunctionToKernels[FF])
+          KernelsToUpdate.insert(K);
+    }
+
+    // Update or attach "calls-indirectly" attribute to those kernels
+    // indicating that they use virtual functions set 'Set'
+    for (Function *K : KernelsToUpdate) {
+      if (!K->hasFnAttribute("calls-indirectly"))
+        K->addFnAttr("calls-indirectly", Set);
+      else {
+        StringRef UsedSets =
+            K->getFnAttribute("calls-indirectly").getValueAsString();
+        if (UsedSets.contains(Set))
+          continue;
+
+        K->removeFnAttr("calls-indirectly");
+        SmallString<64> NewAttr = UsedSets;
```

Comment:
sorry, do you mind explaining exactly what the value of the `calls-indirectly` attr is when the user uses the property? from the sycl property implementation it seems it's set [here](https://github.com/intel/llvm/blob/ff7cf1623b7d7f88eff8ad72123a7a6f36e2c951/sycl/include/sycl/ext/oneapi/experimental/virtual_functions.hpp#L52) but i got lost trying to figure out actually what the value will be set to

Reply:
> do you mind explaining exactly what the value of the calls-indirectly attr is when the user uses the property?

Sure. The value is essentially a string of comma-separate set identifiers where "set identifier" is not exactly defined, i.e. it could be any non-empty set of characters (except comma and maybe null terminator).

From users point of view they can specify which set a function belongs to (through a typename similar to how kernels are named) and then specify which sets of virtual functions are used by a kernel. Typenames are not very friendly to our compiler internals, so we convert them into strings using unique stable name.

Here we don't really care for the content of the attribute (except for maintaining no duplicates of set names), but only care for a fact that we need to add it to a function.

This piece of code handles the following situation:

```c++

q.single_task<Cosntruct>([=]() {
  new (StoragePtr) SomeClass;
}).wait();

// assume_indirect_calls = calls_indirectly<void>
// Headers still need to be updated after a recent spec change
q.single_task<Use>(properties{assume_indirect_calls}, [=]() {
  Storage->virtual_func();
}).wait;
```

In order to execute this code correctly, both `Construct` and `Use` should be present in the same device image which may not always be the case by default depending on device code split. We do not require user to annotate `Construct` kernels (because strictly speaking they don't do virtual calls). But to ensure that they are linked to `Use` kernels, we implicitly record that they use *all* sets of virtual functions which are referenced by vrtable of a constructed object.

Reply:
Thanks, I see, so you can produce "void" as the template argument. If that is done everywhere, I assume that means there is only one virtual function set globally? Is the benefit to allowing custom types there that the user can try to optimize size the of devices images will be loaded and JIT-compiled at runtime (as well as deal with code that may not run on the gpu)?

Reply:
> Thanks, I see, so you can have "void" as the template argument. If that is done everywhere, I assume that means there is only one virtual function set globally?

Right

> Is the benefit to allowing custom types there that the user can try to optimize size the of devices images that will be loaded and JIT-compiled at runtime (as well as deal with code that may not run on the gpu)?

At least two reasons, I think. 1st and most user-visible is handling of optional kernel features. If any virtual function in a set uses `fp64`, then the whole set is considered to be using `fp64`. If you have sub-trees in your class hierarchy which does not use `fp64` and you have corresponding kernels to be able to run your apps on devices without `fp64`, then you will need to explicitly split those virtual functions into two sets.

2nd relies on knowing implementation details, i.e. by knowing that there is runtime linking involved you can carefully craft those sets to either get everything linked in advance, or to avoid any extra (i.e. potentially unused) virtual functions to be linked at all if no kernels are expected to be launched in your application every time.

Reply:
Got it, thanks for the in-depth explanation.